api/ResponseTypes.ts

```diff
@@ -0,0 +1,122 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+export interface Search<T = any> {
+  took: number;
+  timed_out: boolean;
+  _scroll_id?: string;
+  _shards: Shards;
+  hits: {
+    total: {
+      value: number;
+      relation: string;
+    };
```

Comment:
This type is only correct when the server is running Elasticsearch 7.0 and above, and only when the request did not set the property `rest_total_hits_as_int`:

```
rest_total_hits_as_int
    (Optional, boolean) Indicates whether hits.total should be rendered as an integer or an object in the rest search response. Defaults to false.
```

Reply:
For now, this type definitions will land only for `7.x` and above, so it shouldn't be an issue.

> and only when the request did not set the property rest_total_hits_as_int:

How would you solve this? The easiest solutions that come to my mind is to use a union, in the same way I did in `BulkItems`.

Reply:
I think that would create extra work for the 80% case. If possible, I would suggest making the request object into a parameter on the type definition. I've made an example here for how I would write these types:

https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgEJwM4QMoTlBACwCUIMAHAexC2QG8AoZZ5OAczagjbjGGowB+AFzIAriADWISgHcQAbgYBfBgzABPcilz4ipCgIgAeUgEcxZMMgC84qTPkA+W8nOWM1iAA9IIACYY9MhcngD6YJRgcAA2YYTAYBhhmGGgYKJgUJbIysiCTCzoWLoEJGRUNCgAZPSFLCwJSaIgYgC2AEbQ9bnIosU4eGUGlbS1jA0NTRiidMgAbrGWLe1dUAohEDG8-CCiAORskPvIAD7I+xBm+6oNykoMCALWoaMQAHKr0KKl+hVGxjmoTAESisXiiWSqXSmWyKGULjsE0akNEAFYADQ9dicbg7ASzZRY+5qJ40F7-KoAeQ6ACsIAgMshfuVDFVAZtwpFonFpilkjDkPAYrQEa5kcxprMFksIOiMZttnxqAcricidiOFweMqaITiQ8yZ5OW8afTGQAmH5DP5s2hInpS4KLGLLZCYxX4vYXNW5LENHHar0zegakkMIA

Reply:
Also, just for general awareness, I'd recommend looking at a [`discriminated union`](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) feature.

Anyway, what looks simpler to me is just using two separate overloads:
```ts
// Don't export this one, because it is likely to be removed when the legacy interface is removed
interface BasicSearchResponse {
    /* all stable fields */
}

export interface SearchResponse extends BasicSearchResponse {
    hits: {
        total: {
            value: number,
            relation: "gte" | "eq" // @delvedor I suggest using an explicit string literal union here instead of just a `string`
        }
    }
}

// legacy, thus the interface name is verbose:
export interface SearchResponseWithTotalHintsAsInt extends BasicSearchResponse {
    hits: { total: number }
}

// don't export it for the same reason as BasicSearchResponse
interface BasicSearchParams { /* all fields except for rest_total_hits_as_int */ }

export interface SearchParams extends BasicSearchParams {
    rest_total_hits_as_int?: false;
}

export interface SearchParamsWithTotalHintsAsInt extends BasicSearchParams {
    rest_total_hits_as_int: true;
}

export class Client {
    search(params: SearchParams): Promise<SearchResponse>;
    // @deprecated: please use other overload
    search(params: SearchParamsWithTotalHintsAsInt): Promise<SearchResponseWithTotalHintsAsInt>;
}

```